tests/test_notify_v2.py

```diff
@@ -0,0 +1,415 @@
+                'payload_form_field': "baz",
+                'jmespath_query': "not valid JMESPath",
+            }
+            self._put_subscription(doc, replica, codes=requests.codes.unprocessable)
+
+        with self.subTest(f"{replica}, PUT should succceed for valid JMESPath"):
+            sub = {
+                'callback_url': "https://example.com",
+                'method': "POST",
+                'encoding': "application/json",
+                'form_fields': {'foo': "bar"},
+                'payload_form_field': "baz",
+                'jmespath_query': "foo",
+            }
+            self._put_subscription(sub, replica)
+
+        with self.subTest(f"{replica}, PUT should succceed for missing JMESPath"):
+            sub = {
+                'callback_url': "https://example.com",
+                'method': "POST",
+                'encoding': "application/json",
+                'form_fields': {'foo': "bar"},
+                'payload_form_field': "baz",
+            }
+            subscription = self._put_subscription(sub, replica)
+
+        with self.subTest(f"{replica}, GET should succceed"):
+            self._get_subscription(subscription['uuid'], replica)
+
+        with self.subTest(f"{replica}, DELETE should succceed"):
+            self._delete_subscription(subscription['uuid'], replica)
+
+        with self.subTest(f"{replica}, DELETE on non-existent subscription shoudld return not-found"):
+            self._delete_subscription(str(uuid4()), replica, codes=404)
+
+    @testmode.standalone
+    def test_should_notify(self):
+        for replica in Replica:
+            with self.subTest(replica.name):
+                self._test_should_notify(replica)
+
+    def _test_should_notify(self, replica):
+        with self.subTest("notification that does not match document"):
+            sub = self.subscription.copy()
+            sub['jmespath_query'] = 'files."assay.json"[?rna.primer==`george`]'
+            self.assertFalse(notify_v2.should_notify(replica, sub, "CREATE", self.bundle_key))
+
+        with self.subTest("notification with malformed JMESPath query"):
+            sub = self.subscription.copy()
+            sub['jmespath_query'] = 'files."assay.json"[?rna.primer==`george`'
+            self.assertFalse(notify_v2.should_notify(replica, sub, "CREATE", self.bundle_key))
+
+    @testmode.standalone
+    def test_notify(self):
+        with self.subTest("success"):
+            self.assertTrue(notify_v2.notify(self.subscription, "CREATE", self.bundle_key))
+
+        with self.subTest("test multipart/form-data"):
+            sub = self.subscription.copy()
+            sub['encoding'] = "multipart/form-data"
+            self.assertTrue(notify_v2.notify(sub, "CREATE", self.bundle_key))
+
+        with self.subTest("failed delivery"):
+            sub = self.subscription.copy()
+            sub['callback_url'] = f"http://127.0.0.1:{self.app._port}/notification_test_fail"
```

Comment:
Maybe a comment about this failing because the `callback_url` is incorrect?  It wasn't immediately obvious to me.

Reply:
comment changed